### PR TITLE
Support anonymous dimension plotting.

### PR DESCRIPTION
**Pull request by bjlittle**
This PR extends the capability of the `iris.plot` and `iris.quickplot` functions:

- contour
- contourf
- outline
- pcolor
- pcolormesh
- points

to plot cubes with anonymous dimensions. This is done via the `coords` kwarg, which can now take the offset of the anonymous dimension that represents the required plot axis e.g.

```python
>>> print cube
air_temperature / (K)               (latitude: 73; -- : 96)
     Dimension coordinates:
          latitude                           x        -
     Scalar coordinates:
          forecast_period: 6477 hours, bound=(-28083.0, 6477.0) hours
          forecast_reference_time: 1998-03-01 03:00:00
          pressure: 1000.0 hPa
          time: 1998-12-01 00:00:00, bound=(1994-12-01 00:00:00, 1998-12-01 00:00:00)
     Attributes:
          STASH: m01s16i203
          source: Data from Met Office Unified Model
     Cell methods:
          mean within years: time
          mean over years: time
>>> iplt.contourf(cube, coords=(1, 'latitude'))
```
Note that, the axis for the anonymous dimension will be plotted in index space.

### CODE REVIEW

**Comment by marqh** [comment]
Hi @bjlittle 

A primary use case for this appears to be plotting data such as tripolar data, where there are no dim-coords and 2D lon/lat coords.

Please may you comment on the approach in this PR compared to http://scitools.org.uk/iris/docs/latest/examples/General/orca_projection.html as two different ways to achieve a similar aim?

Please may you comment on other use cases for this capability which the example code above does not neatly support?

thank you

**Comment by bjlittle** [comment]
@marqh This PR does not target the case for plotting cubes where only 2d auxiliary (lat/lon) coordinates are available. That was never the intention.

This PR supports the plotting case where there is no coordinate metadata spanning a cube dimension *or* allows the user to nominate the plot axis to be in index space.

File: lib/iris/plot.py [hunk 1]
```diff
@@ -55,21 +55,42 @@
...
+        return ', '.join(result)
+
     def as_coord(coord):
-        coord = cube.coord(coord)
+        if isinstance(coord, int):
+            # Pass through valid dimension indexes.
+            if coord >= ndims:
+                emsg = 'The data dimension ({}) is out of range for ' \
+                    'the dimensionality of the required plot ({})'
+                raise ValueError(emsg.format(coord, ndims))
```
Comment (pelson) [inline lib/iris/plot.py]: Probably an IndexError.

Comment (bjlittle) [inline lib/iris/plot.py]: Yup, probably more helpful. Thanks @pelson

File: lib/iris/plot.py [hunk 2]
```diff
@@ -260,7 +281,12 @@ def _draw_2d_from_bounds(draw_method_name, cube, *args, **kwargs):
         for coord, axis_name, data_dim in zip([u_coord, v_coord],
                                               ['xaxis', 'yaxis'],
                                               [1, 0]):
-            if coord:
+            if coord is None:
```
Comment (pelson) [inline lib/iris/plot.py]: Why do we now need to catch this case?

Comment (pelson) [inline lib/iris/plot.py]: Too late. Just seen it was the `else` below.
I think we can do better to preserve the existing logic and still support `dim=0` :wink

File: lib/iris/plot.py [hunk 3]
```diff
@@ -416,7 +447,7 @@ def _get_plot_objects(args):
                                     (iris.cube.Cube, iris.coords.Coord)):
         # two arguments
         u_object, v_object = args[:2]
-        u, v = _uv_from_u_object_v_object(*args[:2])
```
Comment (pelson) [inline lib/iris/plot.py]: :+1:

File: lib/iris/tests/unit/plot/__init__.py [hunk 4]
```diff
@@ -63,3 +64,61 @@ def assertPointsTickLabels(self, axis, axes=None):
...
+
+@tests.skip_plot
+class MixinCoords(object):
+    """
+    Mixin class of common plotting tests providing 2-dimensional
+    permutations of coordinates and anonymous dimensions.
+
+    """
+    def _check(self, u, v, data=None):
+        self.assertEqual(self.this.call_count, 1)
```
Comment (pelson) [inline lib/iris/tests/unit/plot/__init__.py]: `this` is very JavaScripty...

Comment (bjlittle) [inline lib/iris/tests/unit/plot/__init__.py]: :smile: Happy to change it ...

File: lib/iris/tests/unit/plot/test_contour.py [hunk 5]
```diff
@@ -62,5 +66,20 @@ def test_geoaxes_exception(self):
...
+    def setUp(self):
+        # We have a 2d cube with dimensionality (bar: 3; foo: 4)
+        self.cube = simple_2d(with_bounds=False)
+        self.foo = self.cube.coord('foo').points
+        self.foo_index = np.arange(self.foo.size)
+        self.bar = self.cube.coord('bar').points
+        self.bar_index = np.arange(self.bar.size)
+        self.data = self.cube.data
+        self.dataT = self.data.T
+        self.this = self.patch('matplotlib.pyplot.contour')
```
Comment (pelson) [inline lib/iris/tests/unit/plot/test_contour.py]: Maybe call it `drawing_fn`?

File: lib/iris/tests/unit/plot/test_contour.py [hunk 6]
```diff
@@ -62,5 +66,20 @@ def test_geoaxes_exception(self):
...
+        self.draw = iplt.contour
```
Comment (pelson) [inline lib/iris/tests/unit/plot/test_contour.py]: And this `drawing_fn`.

File: lib/iris/tests/unit/plot/test_contour.py (hunk 5 above)
Comment (bjlittle) [inline lib/iris/tests/unit/plot/test_contour.py]: Yeah, or 'patcher' ... other than 'this'

**Pull request merged**